### PR TITLE
Add support for jaeger style uber-trace-id propagation

### PR DESCRIPTION
**Pull request by nvx**
Extending on the header_type config option, I've added support for jaeger-style uber-trace-id propagation. This implements #22.

I'm not overly familiar with unit testing in Lua however so I haven't made unit tests for the new functionality, but it has been tested manually and seems to work as intended including inter-op with downstream http services using the golang Jaeger library.

As an aside, I noticed setting header_type to eg w3c doesn't prevent the generation of b3 headers to the downstream service if no headers were found in the incoming request. It doesn't stop the configured header_type from being generated as well so not nothing breaks, but did seem a little odd?

### CODE REVIEW

**Comment by seh** [comment]
> I noticed setting header_type to eg w3c doesn't prevent the generation of b3 headers to the downstream service if no headers were found in the incoming request. It doesn't stop the configured header_type from being generated as well so not nothing breaks, but did seem a little odd?

See the discussion starting here: https://github.com/Kong/kong-plugin-zipkin/issues/67#issuecomment-618111142.

**Comment by nvx** [comment]
Cheers for the heads up.

I also just noticed there was a merge conflict due to changes in master that I've just resolved now too. Any ETA on getting this merged @Tieske / @kikito. Happy to do more work on this if anything else is needed.

File: kong/plugins/zipkin/tracing_headers.lua [hunk 1]
```diff
@@ -203,6 +205,57 @@ local function parse_w3c_trace_context_headers(w3c_header)
...
+
+  local trace_id, span_id, parent_id, trace_flags = match(jaeger_header, JAEGER_TRACECONTEXT_PATTERN)
+
+  -- values are not parsable hexidecimal and therefore invalid.
+  if trace_id == nil or span_id == nil or parent_id == nil or trace_flags == nil then
+    warn("invalid jaeger uber-trace-id header; ignoring.")
+  end
+
+  -- valid trace_id is required.
+  if (#trace_id ~= 16 or #trace_id ~= 32) or tonumber(trace_id, 16) == 0 then
```
Comment (greut) [inline kong/plugins/zipkin/tracing_headers.lua:226]:
```suggestion
  if (#trace_id ~= 16 and #trace_id ~= 32) or tonumber(trace_id, 16) == 0 then
```

File: kong/plugins/zipkin/tracing_headers.lua [hunk 2]
```diff
@@ -203,6 +205,57 @@ local function parse_w3c_trace_context_headers(w3c_header)
...
+  end
+
+  -- valid parent_id is required.
+  if #parent_id ~= 16 then
+    warn("invalid jaeger parent ID; ignoring.")
+    return nil, nil, nil, should_sample
+  end
+
+  -- valid flags are required
+  if #trace_flags ~= 1 or #trace_flags ~= 2 then
```
Comment (greut) [inline kong/plugins/zipkin/tracing_headers.lua:244]:
```suggestion
  if #trace_flags ~= 1 and #trace_flags ~= 2 then
```

**Comment by nvx** [comment]
I've accepted those two changes you suggested @greut, resolved the merge conflict with master, and signed the CLA (I assume the CLA is new as I don't remember it flagging up previously).

**Comment by Tieske** [comment]
there still seem to be conflicts, but more importantly it seems to miss tests?

**Comment by nvx** [comment]
GitHub says there are no conflicts with the base branch?

As for the tests, I mentioned in the original post of the PR that I'm not familiar with unit testing in Lua. If there's a quick guide you can point me to for how to run the test suite I'd be more than happy to write up some tests for the new functionality.

**Comment by Tieske** [comment]
try this:
```shell
# install Pongo
git clone https://github.com/kong/kong-pongo
cd kong-pongo
make install
# adjust PATH if you have to here
cd ..

# clone plugin
git clone https://github.com/kong/kong-plugin-zipkin
cd kong-plugin-zipkin

# run tests
pongo run
```
Additional info in the [README.md](https://github.com/Kong/kong-pongo/blob/master/README.md).

**Comment by nvx** [comment]
Unit tests added and a couple of issues fixed.

As an aside, I noticed a couple of issues with parse_w3c_trace_context_headers (which I based the jaeger parse function on) while implementing unit tests for the jaeger headers functionality as they caused my unit tests to fail (the W3C tests pass coincidentally due to other factors).

The first issue is that it returns 4 items on error, but only 3 when no error is returned. I suspect returning just nil 3 times and not also returning should_sample is the correct behaviour as that triggers the should sample logic.

The same function also doesn't return in one of the error cases, but is then caught by the version check just afterwards. I suspect this is just a case of a missing return line: https://github.com/Kong/kong-plugin-zipkin/blob/b3a9807f37937c00407bc4f2df6037cf0cf0e08b/kong/plugins/zipkin/tracing_headers.lua#L168-L170

Both are pretty trivial to fix, but I didn't want to clutter this PR with those changes too.

**Comment by Tieske** [comment]
@kikito mind having a look, since you did most of the maintenance on this repo iirc?

**Comment by kikito** [comment]
Hello, thank you for sending this PR and apologies for the time it took us to merge it. 

I merged this on another pr (#101) after rebasing over master.

> I noticed a couple of issues with parse_w3c_trace_context_headers

Thanks for this. I have done the relevant changes in #100

**Pull request closed without merging**